File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/crypt/CrlCacheFileSystem.kt
package com.jetbrains.signatureverifier.crypt

import org.bouncycastle.cert.X509CRLHolder
import org.jetbrains.annotations.NotNull
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.attribute.BasicFileAttributes
import kotlin.io.path.extension
import kotlin.io.path.pathString
import kotlin.streams.toList

open class CrlCacheFileSystem {
  private val _cacheDir: Path

  constructor() : this("crlscache")

  constructor(@NotNull cacheDir: String) {
    val cacheDirName = cacheDir
    _cacheDir = Paths.get(System.getProperty("java.io.tmpdir"), cacheDirName)
  }

  open fun GetCrls(@NotNull issuerId: String): Collection<X509CRLHolder> {
    val crlFiles = getCrlFiles(issuerId)
    val res = mutableListOf<X509CRLHolder>()
    for (path in crlFiles) {
      val file = path.toFile()
      file.inputStream().use {
        val crl = X509CRLHolder(it)
        res.add(crl)
      }
    }
    return res.toList()
  }

  open fun UpdateCrls(issuerId: String, crlsData: List<ByteArray>) {
    cleanUpCrls(issuerId)
    saveCrls(issuerId, crlsData)
  }

  private fun getCrlFiles(issuerId: String): List<Path> {
    fun filter(path: Path, attrs: BasicFileAttributes): Boolean {
      return path.fileName.toString().startsWith(issuerId) && path.extension == "crl"
    }

    ensureCacheDirectory()
    return Files.find(_cacheDir, 1, ::filter).toList()
  }

  private fun ensureCacheDirectory() {
    if (!Files.exists(_cacheDir))
      Files.createDirectory(_cacheDir)
  }

  private fun cleanUpCrls(issuerId: String) {
    for (crlFile in getCrlFiles(issuerId))
      Files.deleteIfExists(crlFile)
  }

  private fun saveCrls(issuerId: String, crlsData: List<ByteArray>) {
    if (crlsData.count() == 1) {
      val crlFileName = "${issuerId}.crl"
      saveCrl(crlFileName, crlsData[0])
    } else {
      for (i in 0 until crlsData.count()) {
        val crlFileName = "${issuerId}_${i}.crl"
        saveCrl(crlFileName, crlsData[i])
      }
    }
  }

  private fun saveCrl(crlFileName: String, crlData: ByteArray) {
    val crlFilePath = Paths.get(_cacheDir.pathString, crlFileName)
    crlFilePath.toFile().writeBytes(crlData)
  }
}

File: net/JetBrains.FormatRipper/src/Crypt/CrlCacheFileSystem.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using JetBrains.Annotations;
using Org.BouncyCastle.X509;

namespace JetBrains.SignatureVerifier.Crypt
{
  public class CrlCacheFileSystem
  {
    private readonly string _cacheDir;
    private readonly X509CrlParser _crlParser = new();

    public CrlCacheFileSystem()
      : this("crlscache")
    {
    }

    public CrlCacheFileSystem([NotNull] string cacheDir)
    {
      var cacheDirName = cacheDir ?? throw new ArgumentNullException(nameof(cacheDir));
      _cacheDir = Path.Combine(Path.GetTempPath(), cacheDirName);
    }

    public ReadOnlyCollection<X509Crl> GetCrls([NotNull] string issuerId)
    {
      if (issuerId == null) throw new ArgumentNullException(nameof(issuerId));

      var crlFiles = getCrlFiles(issuerId);
      var res = new List<X509Crl>();

      foreach (var path in crlFiles)
      {
        using var stream = File.OpenRead(path);
        var crl = _crlParser.ReadCrl(stream);
        res.Add(crl);
      }

      return res.AsReadOnly();
    }

    public virtual void UpdateCrls(string issuerId, List<byte[]> crlsData)
    {
      cleanUpCrls(issuerId);
      saveCrls(issuerId, crlsData);
    }

    private IEnumerable<string> getCrlFiles(string issuerId)
    {
      ensureCacheDirectory();
      return Directory.EnumerateFiles(_cacheDir, $"{issuerId}*.crl", SearchOption.TopDirectoryOnly);
    }

    private void ensureCacheDirectory()
    {
      if (!Directory.Exists(_cacheDir))
        Directory.CreateDirectory(_cacheDir);
    }

    private void cleanUpCrls(string issuerId)
    {
      foreach (var crlFile in getCrlFiles(issuerId))
        File.Delete(crlFile);
    }

    private void saveCrls(string issuerId, List<byte[]> crlsData)
    {
      if (crlsData.Count == 1)
      {
        var crlFileName = $"{issuerId}.crl";
        saveCrl(crlFileName, crlsData[0]);
      }
      else
      {
        for (var i = 0; i < crlsData.Count; i++)
        {
          var crlFileName = $"{issuerId}_{i}.crl";
          saveCrl(crlFileName, crlsData[i]);
        }
      }
    }

    private void saveCrl(string crlFileName, byte[] crlData)
    {
      var crlFilePath = Path.Combine(_cacheDir, crlFileName);
      File.WriteAllBytes(crlFilePath, crlData);
    }
  }
}